keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2314 @@
+    - Implementation Phase 1:
+        - kubelet
+        - kube-apiserver
+        - kube-scheduler
+    - Future phases (with quota and limitranges support):
+        - kubelet
+        - kube-apiserver
+        - kube-scheduler
+        - kube-controller-manager
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+
+###### Does enabling the feature change any default behavior?
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes it can.
+
+Running workloads continue to work without any changes. Restarting or
+re-deploying a workload causes it to fail as the requested QoS-class resources
+are not available in Kubernetes anymore. The resources are still supported by
+the underlying runtime but disabling the feature in Kubernetes makes them
+unavailable and the related PodSpec fields are not accepted in validation.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Workloads might have failed because of unsupported fields in the pod spec and
+need to be restarted.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+Kubelet unit tests are extended to verify that no QoS-class resource
+assignments are correctly passed down to the CRI API, also verifying that
+assignments are not passed down if the feature gate is disabled.
+
+Apiserver unit tests are extended to verify that the new fields in PodSpec are
+preserved over updates of the Pod object, even when the feature gate is
+disabled.
+
```

Comment:
Scheduler probably needs tests as well.

Reply:
Definitely, node fitting must be covered. Thanks for noting. Updated.